xds/internal/balancer/gracefulswitch/gracefulswitch.go

```diff
@@ -0,0 +1,399 @@
+	// guaranteed grpc will call the balancer API synchronously, and for the
+	// pending to also be closed would require another update from grpc. The
+	// current being closed is not a problem for this ResolverError() call, as
+	// in the scenario that the current balancer can be closed (current +
+	// pending balancers are both populated), this error will always be
+	// forwarded to the pending. Thus, there is a guarantee that this will not
+	// break the balancer API of the balancer by updating after closing.
+	balToUpdate.ResolverError(err)
+}
+
+// UpdateSubConnState forwards the update to the appropriate child.
+func (gsb *Balancer) UpdateSubConnState(sc balancer.SubConn, state balancer.SubConnState) {
+	// At any given time, the current balancer may be closed in the situation
+	// where there is a current balancer and a pending balancer. Thus, the
+	// swap() (close() + clearing of SubConns) operation is also guarded by this
+	// currentMu. Whether the current has been deleted or not will be checked
+	// from reading from the scToSubBalancer map, as that will be cleared in the
+	// atomic swap() operation. We are guaranteed that there can only be one
+	// balancer closed within this UpdateSubConnState() call, as multiple
+	// balancers closed would require another call to SwitchTo() which we are
+	// guaranteed won't be called concurrently. UpdateSubConnState() is
+	// different to UpdateClientConnState() and ResolverError(), as for those
+	// functions you don't update the current balancer in the situation where
+	// the current balancer could be closed (current + pending balancer
+	// populated).
+	gsb.currentMu.Lock()
+	defer gsb.currentMu.Unlock()
+	gsb.mu.Lock()
+	if gsb.closed {
+		gsb.mu.Unlock()
+		return
+	}
+	// This SubConn update will forward to the current balancer even if there is
+	// a pending present. This is because if this balancer does not forward the
+	// update, the picker from the current will not be updated with SubConns,
+	// leading to the possibility that the ClientConn constantly picks bad
+	// SubConns in the Graceful Switch period.
+	var balToUpdate balancer.Balancer
+	if gsb.balancerCurrent != nil && gsb.balancerCurrent.subconns[sc] {
+		balToUpdate = gsb.balancerCurrent
+	}
+	if gsb.balancerPending != nil && gsb.balancerPending.subconns[sc] {
+		balToUpdate = gsb.balancerPending
+	}
+	if balToUpdate == nil {
+		// SubConn belonged to a stale LB policy that has not yet fully closed.
+		gsb.mu.Unlock()
+		return
+	}
+	gsb.mu.Unlock()
+	balToUpdate.UpdateSubConnState(sc, state)
+}
+
+// Close closes any active child balancers.
+func (gsb *Balancer) Close() {
+	gsb.mu.Lock()
+	gsb.closed = true
+	currentBalancerToClose := gsb.balancerCurrent
+	gsb.balancerCurrent = nil
+	pendingBalancerToClose := gsb.balancerPending
+	gsb.balancerPending = nil
+	gsb.mu.Unlock()
+
+	if currentBalancerToClose != nil {
+		currentBalancerToClose.Close()
+	}
+	if pendingBalancerToClose != nil {
+		pendingBalancerToClose.Close()
+	}
```

Comment:
Optional: we do a lot of `nil` checks before calling `Close`.  Instead, unconditionally call `Close` and put the nil check at the start of `balancerWrapper.Close`.

Reply:
> Plus, I don't really get how to move the nil check to balancerWrapper

```go
func (bw *balancerWrapper) Close() {
	if bw == nil {
		return
	}
	....
}
```

That's it!

It's fine either way, but a little cleaner (less clutter) with the `nil` check done in bw, at the tradeoff of slightly more magic (the reader might thing "hey, there's a missing nil check here!").

This is similar to handling a base case at the entry of the function instead of before you recurse:

```go
func recurse(n *treeNode) {
	if n == nil {
		return
	}
	recurse(n.left)
	recurse(n.right)

	/// vs...

	if n.left != nil {
		recurse(n.left)
	}
	if n.right != nil {
		recurse(n.right)
	}
}
```

Reply:
Switched.